File: backend/incident_engine.py
"""End-to-end incident report: one attack, start to finish.

The rest of the app shows the loop in pieces — the Lab plans and scores, the
cockpit compares detectors, Defend retrains. This composes all of it into a
single document about a single incident, so the whole causal chain is
readable in one place:

    who was targeted  ->  why this attack  ->  what was sent
                      ->  what each detector said
                      ->  what got through and why
                      ->  what the model learned from the misses

Every phase is computed in the same call against the same run, so the
narrative cannot drift out of sync with the numbers the way three separate
pages can.

Pure stdlib — the serverless bundle ships only fastapi+pydantic.
"""

from __future__ import annotations

import random
from datetime import datetime, timedelta, timezone
from typing import Any, Optional

import cockpit_engine
import defender_engine
import lab_engine
from risk import BLOCK_THRESHOLD, REVIEW_THRESHOLD


def _phase(id_, title, headline, detail, facts=None):
    return {
        "id": id_,
        "title": title,
        "headline": headline,
        "detail": detail,
        "facts": facts or [],
    }


def build_report(
    target_id: str = "C0001",
    vector: Optional[str] = None,
    attack_type: Optional[str] = None,
    seed: Optional[int] = None,
    include_retraining: bool = True,
) -> dict[str, Any]:
    effective_seed = seed if seed is not None else random.SystemRandom().randrange(2**31)
    rng = random.Random(f"incident:{target_id}:{effective_seed}")

    customer = lab_engine.get_customer(target_id)
    family = attack_type or cockpit_engine.VECTOR_FAMILIES.get(vector or "", None)

    # ---- Phase 1-2: profile and plan ------------------------------------
    plan = lab_engine.plan_attack(customer, attack_type=family)
    family = plan["attack_type"]
    ranked = sorted(plan["candidates"].items(), key=lambda kv: -kv[1])

    # ---- Phase 3: synthesis ---------------------------------------------
    records = lab_engine.generate_and_score(
        customer, family, rng, difficulty="medium", intensity=0.7,
    )

    # ---- Phase 4: scoring, both detectors -------------------------------
    started = datetime.now(timezone.utc)
    payments = []
    for index, record in enumerate(records):
        features = record["features"]
        legacy = cockpit_engine.legacy_score(features)
        payments.append({
            "step": index + 1,
            "id": record["id"],
            "at": (started + timedelta(minutes=7 * index)).isoformat(timespec="seconds"),
            "amount": record["amount"],
            "amountRatio": round(record["amount"] / features["amount_baseline"], 2),
            "hour": record["hour"],
            "velocity": features["txn_velocity_1h"],
            "isNewPayee": features["is_new_payee"],
            "isInternational": features["is_international"],
            "isNewDevice": features["is_new_device"],
            "signal": record["signal"],
            "riskScore": record["risk_score"],
            "action": record["recommended_action"],
            "flagged": record["status"] == "flagged",
            "reasons": record["risk_reasons"],
            "contributions": record["contributions"],
            "explanation": record["explanation"],
            "legacyFlagged": legacy["flagged"],
            "legacyReasons": legacy["reasons"],
        })

    total = len(payments)
    hardened_caught = sum(1 for p in payments if p["flagged"])
    legacy_caught = sum(1 for p in payments if p["legacyFlagged"])
    evaded = [p for p in payments if not p["flagged"]]
    blocked = sum(1 for p in payments if p["action"] == "BLOCK")
    stepped_up = sum(1 for p in payments if p["action"] == "STEP_UP_AUTH")

    value_total = round(sum(p["amount"] for p in payments), 2)
    value_stopped = round(sum(p["amount"] for p in payments if p["flagged"]), 2)
    value_through = round(sum(p["amount"] for p in evaded), 2)
    value_legacy_through = round(sum(p["amount"] for p in payments if not p["legacyFlagged"]), 2)

    # ---- Phase 5: why the misses got through ----------------------------
    if evaded:
        avg_ratio = sum(p["amountRatio"] for p in evaded) / len(evaded)
        avg_velocity = sum(p["velocity"] for p in evaded) / len(evaded)
        with_payee = sum(1 for p in evaded if p["isNewPayee"])
        evasion_detail = (
            f"The {len(evaded)} payments that got through averaged {avg_ratio:.2f}x this "
            f"account's baseline at {avg_velocity:.1f} payments/hour, and {with_payee} of "
            f"{len(evaded)} went to a new payee. None accumulated enough signal to reach the "
            f"{REVIEW_THRESHOLD:.2f} review line — each one is individually unremarkable, "
            f"which is the point of the technique."
        )
    else:
        evasion_detail = (
            "Every payment in this sequence crossed the review threshold. Raise the "
            "difficulty to force the attacker into stealthier pacing."
        )

    # ---- Phase 6: what the defender learned -----------------------------
    retraining = defender_engine.train(seed=effective_seed) if include_retraining else None

    phases = [
        _phase(
            "profile", "01 · Target profiled",
            f"{customer.get('name', customer['customer_id'])} in {customer['city']}",
            f"The attacker starts from observable behaviour, not guesswork. This account "
            f"averages ${customer['average_amount']:,} across about {customer['daily_txns']} "
            f"payment(s) a day, on a device it uses {customer['device_stability']:.0%} of the "
            f"time, with spending regularity {customer['spending_regularity']}. Those three "
            f"numbers decide which attack works.",
            [
                {"label": "Account", "value": f"{customer['customer_id']} · {customer['city']}"},
                {"label": "Baseline payment", "value": f"${customer['average_amount']:,}"},
                {"label": "Usual cadence", "value": f"{customer['daily_txns']}/day"},
                {"label": "Device stability", "value": f"{customer['device_stability']}"},
                {"label": "Spending regularity", "value": f"{customer['spending_regularity']}"},
            ],
        ),
        _phase(
            "plan", "02 · Attack selected",
            f"{family.replace('_', ' ').title()} scored highest",
            f"Each attack family is ranked against this specific account's weak points. "
            f"{family.replace('_', ' ').title()} won at {plan['candidates'].get(family, 0)} "
            f"because the account's softest surface is "
            f"{plan['primary_weakness'].replace('_', ' ')}. "
            f"A different account would get a different plan.",
            [{"label": name.replace("_", " ").title(), "value": f"{score}"} for name, score in ranked[:5]],
        ),
        _phase(
            "synthesis", "03 · Payload synthesized",
            f"{total} payments over {total * 7} minutes",
            f"The generator shapes a sequence rather than a single payment: amounts ramp "
            f"between {min(p['amountRatio'] for p in payments)}x and "
            f"{max(p['amountRatio'] for p in payments)}x the account baseline, velocity moves "
            f"between {min(p['velocity'] for p in payments)} and "
            f"{max(p['velocity'] for p in payments)} per hour, worth ${value_total:,.2f} in total.",
            [
                {"label": "Payments", "value": str(total)},
                {"label": "Total value", "value": f"${value_total:,.2f}"},
                {"label": "Amount range", "value": f"{min(p['amountRatio'] for p in payments)}x – {max(p['amountRatio'] for p in payments)}x"},
                {"label": "Modality", "value": plan["modality"]},
            ],
        ),
        _phase(
            "scoring", "04 · Both detectors scored it",
            f"Legacy caught {legacy_caught}, hardened caught {hardened_caught}, of {total}",
            f"Every payment was scored twice: once by flat rules that know nothing about this "
            f"account, and once by the hardened scorer that grades each signal against the "
            f"account's own baseline. Of {total} payments the hardened scorer blocked "
            f"{blocked} outright and sent {stepped_up} to step-up authentication.",
            [
                {"label": "Legacy caught", "value": f"{legacy_caught}/{total}"},
                {"label": "Hardened caught", "value": f"{hardened_caught}/{total}"},
                {"label": "Blocked", "value": str(blocked)},
                {"label": "Step-up", "value": str(stepped_up)},
                {"label": "Value stopped", "value": f"${value_stopped:,.2f}"},
                {"label": "Value through", "value": f"${value_through:,.2f}"},
            ],
        ),
        _phase(
            "evasion", "05 · What got through",
            f"{len(evaded)} of {total} payments evaded detection",
            evasion_detail,
            [
                {"label": "Evaded", "value": f"{len(evaded)}/{total}"},
                {"label": "Value through (hardened)", "value": f"${value_through:,.2f}"},
                {"label": "Value through (legacy)", "value": f"${value_legacy_through:,.2f}"},
            ],
        ),
    ]

    if retraining:
        rounds = retraining["rounds"]
        phases.append(_phase(
            "hardening", "06 · Defender retrained on the misses",
            f"Recall {rounds[0]['recall']:.0%} → {rounds[-1]['recall']:.0%} over "
            f"{len(rounds)} rounds",
            f"The evaded payments become training data. Each round mines what the previous "
            f"model missed and retrains on it, which is why recall climbs while the test "
            f"split stays fixed. Precision moves from {rounds[0]['precision']:.0%} to "
            f"{rounds[-1]['precision']:.0%} — catching stealthier fraud costs some precision, "
            f"and that trade is visible rather than hidden.",
            [
                {"label": f"Round {r['round']} recall", "value": f"{r['recall']:.1%}"}
                for r in rounds
            ] + [
                {"label": "AUC", "value": f"{rounds[0]['auc']:.3f} → {rounds[-1]['auc']:.3f}"},
            ],
        ))

    return {
        "incidentId": f"INC-{customer['customer_id']}-{effective_seed % 100000:05d}",
        "generatedAt": started.isoformat(timespec="seconds"),
        "seed": effective_seed,
        "target": customer,
        "vector": vector,
        "attack": {
            "family": family,
            "modality": plan["modality"],
            "primaryWeakness": plan["primary_weakness"],
            "description": plan["description"],
            "candidates": plan["candidates"],
            "lure": cockpit_engine.VECTOR_LURES.get(
                vector or "", cockpit_engine.VECTOR_LURES["bec-email"]
            ).format(
                name=customer.get("name") or customer["customer_id"],
                amount=max(p["amount"] for p in payments),
            ),
        },
        "phases": phases,
        "payments": payments,
        "summary": {
            "total": total,
            "hardenedCaught": hardened_caught,
            "legacyCaught": legacy_caught,
            "evaded": len(evaded),
            "blocked": blocked,
            "steppedUp": stepped_up,
            "detectionRate": round(hardened_caught / total, 4) if total else 0.0,
            "legacyDetectionRate": round(legacy_caught / total, 4) if total else 0.0,
            "evasionRate": round(len(evaded) / total, 4) if total else 0.0,
            "valueTotal": value_total,
            "valueStopped": value_stopped,
            "valueThrough": value_through,
            "valueThroughLegacy": value_legacy_through,
        },
        "retraining": retraining,
        "thresholds": {"review": REVIEW_THRESHOLD, "block": BLOCK_THRESHOLD},
        "provenance": {
            "generator": "lab_engine attack families",
            "scorers": "legacy static rules vs risk.score_transaction",
            "retraining": "defender_engine, 3-round adversarial loop" if retraining else None,
            "syntheticOnly": True,
        },
    }

File: lib/incident-engine.js
/**
 * End-to-end incident report.
 *
 * The rest of the app shows the loop in pieces. This composes all of it into
 * one document about one incident, computed in a single pass, so the narrative
 * cannot drift out of sync with the numbers the way separate pages can.
 */

import { makeRng, randomSeed } from "./rng.js";
import { getCustomer, planAttack, generateAndScore, summarize } from "./lab-engine.js";
import { legacyScore, REVIEW_THRESHOLD, BLOCK_THRESHOLD } from "./risk.js";
import { train } from "./benchmark-engine.js";
import { labelFor } from "./taxonomy.js";

const LURES = {
  vishing: (n, a) =>
    `Caller, using a cloned voice of the CFO: "${n}, I'm in a vendor board meeting. Please authorise the $${a} supplier invoice to Apex Meridian Global before the cutoff."`,
  video_deepfake: (n, a) =>
    `A face-swapped video call stands in for ${n} during step-up identity verification, authorising a $${a} transfer to a newly registered beneficiary.`,
  synthetic_identity: (n, a) =>
    `An applicant profile blends a genuine identity number with synthesised biometrics, opening an account that immediately moves $${a}.`,
  phishing: (n, a) =>
    `A reply is injected into a live supplier invoice thread: "Please note our updated remittance details. Kindly settle the outstanding $${a} to the account below."`,
  behavioral_drift: (n, a) =>
    `An autonomous storefront with generated listings and reviews takes a $${a} order it will never fulfil.`,
  account_takeover: (n, a) =>
    `Breached credentials and a spoofed device open a session as ${n}, then move $${a} to a beneficiary added minutes earlier.`,
  device_switch: (n, a) =>
    `A payment for $${a} arrives from a device and browser fingerprint this account has never used.`,
  velocity_anomaly: (n, a) =>
    `A burst of payments peaking at $${a} is pushed through inside one hour, each one individually unremarkable.`,
  sleeper_transaction_pacing: (n, a) =>
    `Small payments warm the account's baseline over several days before a $${a} cash-out.`,
  adversarial_probing: (n, a) =>
    `Feature values are nudged one at a time around a $${a} payment, hunting for the exact point the detector stops firing.`,
};

const fmt = (n) => Math.round(n).toLocaleString("en-US");

export function buildReport({ targetId = "C0001", attackType = null, seed = null, includeTraining = true } = {}) {
  const effectiveSeed = seed ?? randomSeed();
  const customer = getCustomer(targetId);
  const rng = makeRng(`incident:${customer.id}:${effectiveSeed}`);

  const plan = planAttack(customer, attackType);
  const raw = generateAndScore(customer, plan.attackType, rng, "medium", 0.7);
  const defence = summarize(raw);
  const started = new Date();

  const payments = raw.map((r, i) => {
    const legacy = legacyScore(r.features);
    return {
      ...r,
      at: new Date(started.getTime() + i * 7 * 60000).toISOString(),
      legacyFlagged: legacy.flagged,
      legacyReasons: legacy.reasons,
    };
  });

  const total = payments.length;
  const hardenedCaught = payments.filter((p) => p.flagged).length;
  const legacyCaught = payments.filter((p) => p.legacyFlagged).length;
  const evaded = payments.filter((p) => !p.flagged);
  const valueTotal = payments.reduce((a, p) => a + p.amount, 0);
  const valueStopped = payments.filter((p) => p.flagged).reduce((a, p) => a + p.amount, 0);
  const valueThroughLegacy = payments.filter((p) => !p.legacyFlagged).reduce((a, p) => a + p.amount, 0);

  const ranked = Object.entries(plan.candidates).sort((a, b) => b[1] - a[1]);
  const training = includeTraining ? train(effectiveSeed) : null;

  const evasionDetail = evaded.length
    ? `The ${evaded.length} payments that got through averaged ${(
        evaded.reduce((a, p) => a + p.amountRatio, 0) / evaded.length
      ).toFixed(2)}x this account's baseline at ${(
        evaded.reduce((a, p) => a + p.velocity, 0) / evaded.length
      ).toFixed(1)} payments per hour, and ${evaded.filter((p) => p.isNewPayee).length} of ${
        evaded.length
      } went to a new payee. None accumulated enough signal to reach the ${REVIEW_THRESHOLD.toFixed(
        2
      )} review line. Each one is individually unremarkable, which is the point of the technique.`
    : "Every payment in this sequence crossed the review threshold. Raise the difficulty to force stealthier pacing.";

  const phases = [
    {
      id: "profile",
      title: "Target profiled",
      headline: `${customer.name} in ${customer.city}`,
      detail: `The attacker starts from observable behaviour, not guesswork. This account averages $${fmt(customer.baseline)} across about ${customer.daily} payment(s) a day, on a device it uses ${Math.round(customer.deviceStability * 100)}% of the time, with spending regularity ${customer.regularity}. Those three numbers decide which attack works.`,
      facts: [
        { label: "Account", value: `${customer.id} · ${customer.city}` },
        { label: "Baseline payment", value: `$${fmt(customer.baseline)}` },
        { label: "Usual cadence", value: `${customer.daily}/day` },
        { label: "Device stability", value: String(customer.deviceStability) },
        { label: "Spending regularity", value: String(customer.regularity) },
      ],
    },
    {
      id: "plan",
      title: "Attack selected",
      headline: `${plan.label} scored highest`,
      detail: `Every attack family is ranked against this specific account's weak points. ${plan.label} won at ${plan.candidates[plan.attackType]} because the softest surface is ${plan.primaryWeakness}. A different account gets a different plan.`,
      facts: ranked.slice(0, 5).map(([k, v]) => ({
        label: labelFor(k),
        value: String(v),
      })),
    },
    {
      id: "synthesis",
      title: "Payload synthesised",
      headline: `${total} payments over ${total * 7} minutes`,
      detail: `The generator shapes a sequence rather than a single payment. Amounts ramp between ${Math.min(...payments.map((p) => p.amountRatio))}x and ${Math.max(...payments.map((p) => p.amountRatio))}x the account baseline, velocity moves between ${Math.min(...payments.map((p) => p.velocity))} and ${Math.max(...payments.map((p) => p.velocity))} per hour, worth $${fmt(valueTotal)} in total.`,
      facts: [
        { label: "Payments", value: String(total) },
        { label: "Total value", value: `$${fmt(valueTotal)}` },
        { label: "Modality", value: plan.modality },
      ],
    },
    {
      id: "scoring",
      title: "Both detectors scored it",
      headline: `Legacy caught ${legacyCaught}, hardened caught ${hardenedCaught}, of ${total}`,
      detail: `Every payment was scored twice: once by flat rules that know nothing about this account, and once by a scorer that grades each signal against the account's own baseline. The hardened scorer blocked ${defence.blocked} outright and sent ${defence.steppedUp} to step-up authentication.`,
      facts: [
        { label: "Legacy caught", value: `${legacyCaught}/${total}` },
        { label: "Hardened caught", value: `${hardenedCaught}/${total}` },
        { label: "Blocked", value: String(defence.blocked) },
        { label: "Step-up", value: String(defence.steppedUp) },
        { label: "Value stopped", value: `$${fmt(valueStopped)}` },
      ],
    },
    {
      id: "evasion",
      title: "What got through",
      headline: `${evaded.length} of ${total} payments evaded detection`,
      detail: evasionDetail,
      facts: [
        { label: "Evaded", value: `${evaded.length}/${total}` },
        { label: "Value through, hardened", value: `$${fmt(valueTotal - valueStopped)}` },
        { label: "Value through, legacy", value: `$${fmt(valueThroughLegacy)}` },
      ],
    },
  ];

  if (training) {
    const r = training.rounds;
    phases.push({
      id: "hardening",
      title: "Defender retrained on the misses",
      headline: `Recall ${Math.round(r[0].recall * 100)}% to ${Math.round(r[r.length - 1].recall * 100)}% over ${r.length} rounds`,
      detail: `The evaded payments become training data. Each round mines what the previous model missed and retrains on it, which is why recall climbs while the test split stays fixed. Precision moves from ${Math.round(r[0].precision * 100)}% to ${Math.round(r[r.length - 1].precision * 100)}%. Catching stealthier fraud costs some precision, and that trade is visible rather than hidden.`,
      facts: r
        .map((x) => ({ label: `${x.name} recall`, value: `${(x.recall * 100).toFixed(1)}%` }))
        .concat([{ label: "AUC", value: `${r[0].auc.toFixed(3)} to ${r[r.length - 1].auc.toFixed(3)}` }]),
    });
  }

  const peak = Math.max(...payments.map((p) => p.amount));

  return {
    incidentId: `INC-${customer.id}-${String(effectiveSeed % 100000).padStart(5, "0")}`,
    generatedAt: started.toISOString(),
    seed: effectiveSeed,
    target: customer,
    attack: {
      family: plan.attackType,
      label: plan.label,
      modality: plan.modality,
      primaryWeakness: plan.primaryWeakness,
      candidates: plan.candidates,
      lure: (LURES[plan.attackType] || LURES.phishing)(customer.name, fmt(peak)),
    },
    phases,
    payments,
    summary: {
      total,
      hardenedCaught,
      legacyCaught,
      evaded: evaded.length,
      blocked: defence.blocked,
      steppedUp: defence.steppedUp,
      detectionRate: Number((hardenedCaught / total).toFixed(4)),
      legacyDetectionRate: Number((legacyCaught / total).toFixed(4)),
      evasionRate: Number((evaded.length / total).toFixed(4)),
      valueTotal: Number(valueTotal.toFixed(2)),
      valueStopped: Number(valueStopped.toFixed(2)),
      valueThrough: Number((valueTotal - valueStopped).toFixed(2)),
      valueThroughLegacy: Number(valueThroughLegacy.toFixed(2)),
    },
    training,
    thresholds: { review: REVIEW_THRESHOLD, block: BLOCK_THRESHOLD },
    provenance: {
      generator: "lab-engine attack families",
      scorers: "legacy static rules vs per-account graded scorer",
      training: training ? "3-round adversarial loop, logistic regression" : null,
      syntheticOnly: true,
    },
  };
}
